Origin: uniontech
Reviewed-By: zhangfeilong@uniontech.com
Last-Update: 2021-02-25

--- libsigsegv-2.12.orig/configure.ac
+++ libsigsegv-2.12/configure.ac
@@ -211,6 +211,13 @@ SV_TRY_FAULT([OSF/1 Alpha], sv_cv_fault_
   [scp->sc_traparg_a0])
 
 dnl FIXME: Put in some more known values into the third argument.
+SV_TRY_FAULT([Linux/sw_64], sv_cv_fault_linux_sw_64,
+  [sw_64*-*-linux2[4-9]* | sw_64*-*-linux2.[4-9]*],
+  [],
+  [int sig, int code, struct sigcontext *scp],
+  [scp->sc_traparg_a0])
+
+dnl FIXME: Put in some more known values into the third argument.
 SV_TRY_FAULT([NetBSD Alpha], sv_cv_fault_netbsd_alpha,
   [alpha*-*-osf[4-9]* | alpha-*-*bsd*],
   [#include "$srcdir/src/fault-netbsd-alpha.c"],
@@ -283,6 +290,13 @@ if test -z "$CFG_FAULT" && test "$sv_cv_
   esac
   FAULT_CONTEXT='struct sigcontext'
 fi
+if test -z "$CFG_FAULT" && test "$sv_cv_fault_linux_sw_64" = yes; then
+  case "$host_cpu" in
+    sw_64*) CFG_FAULT=fault-linux-sw_64.h ;;
+    *) CFG_FAULT=fault-linux.h ;;
+  esac
+  FAULT_CONTEXT='struct sigcontext'
+fi
 if test -z "$CFG_FAULT" && test "$sv_cv_fault_osf_alpha" = yes; then
   case "$host_cpu" in
     alpha*) CFG_FAULT=fault-osf-alpha.h ;;
@@ -404,6 +418,9 @@ if test -z "$CFG_FAULT" && test "$sv_cv_
             alpha*)
               CFG_FAULT=fault-linux-alpha.h
               ;;
+            sw_64*)
+              CFG_FAULT=fault-linux-sw_64.h
+              ;;
             arm* | strongarm* | xscale* | aarch64*)
               CFG_FAULT=fault-linux-arm.h
               ;;
@@ -573,6 +590,10 @@ else
           CFG_FAULT=fault-linux-alpha-old.h
           FAULT_CONTEXT='struct sigcontext'
           ;;
+        sw_64*)
+          CFG_FAULT=fault-linux-sw_64-old.h
+          FAULT_CONTEXT='struct sigcontext'
+          ;;
         arm* | strongarm* | xscale*)
           CFG_FAULT=fault-linux-arm-old.h
           FAULT_CONTEXT='struct sigcontext'
--- libsigsegv-2.12.orig/src/Makefile.am
+++ libsigsegv-2.12/src/Makefile.am
@@ -34,6 +34,7 @@ noinst_HEADERS = \
   fault-hurd.h fault-hurd-i386.h \
   fault-irix.h fault-irix-mips.h \
   fault-linux.h fault-linux-alpha.h fault-linux-alpha-old.h \
+  fault-linux-sw_64.h fault-linux-sw_64-old.h \
   fault-linux-arm.h fault-linux-arm-old.h \
   fault-linux-cris.h fault-linux-cris-old.h \
   fault-linux-hppa.h fault-linux-hppa-old.h \
--- libsigsegv-2.12.orig/src/Makefile.in
+++ libsigsegv-2.12/src/Makefile.in
@@ -348,6 +348,7 @@ noinst_HEADERS = \
   fault-hurd.h fault-hurd-i386.h \
   fault-irix.h fault-irix-mips.h \
   fault-linux.h fault-linux-alpha.h fault-linux-alpha-old.h \
+  fault-linux-sw_64.h fault-linux-sw_64-old.h \
   fault-linux-arm.h fault-linux-arm-old.h \
   fault-linux-cris.h fault-linux-cris-old.h \
   fault-linux-hppa.h fault-linux-hppa-old.h \
--- /dev/null
+++ libsigsegv-2.12/src/fault-linux-sw_64-old.h
@@ -0,0 +1,22 @@
+/* Fault handler information.  Linux/sw_64 version.
+   Copyright (C) 2002  Bruno Haible <bruno@clisp.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2, or (at your option)
+   any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software Foundation,
+   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */
+
+#include <asm/sigcontext.h>
+
+#define SIGSEGV_FAULT_HANDLER_ARGLIST  int sig, int code, struct sigcontext *scp
+#define SIGSEGV_FAULT_CONTEXT  scp
+#define SIGSEGV_FAULT_STACKPOINTER  scp->sc_regs[30]
--- /dev/null
+++ libsigsegv-2.12/src/fault-linux-sw_64.h
@@ -0,0 +1,20 @@
+/* Fault handler information.  Linux/sw_64 version when it supports POSIX.
+   Copyright (C) 2002, 2009  Bruno Haible <bruno@clisp.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2, or (at your option)
+   any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software Foundation,
+   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */
+
+#include "fault-posix-ucontext.h"
+
+#define SIGSEGV_FAULT_STACKPOINTER  ((ucontext_t *) ucp)->uc_mcontext.sc_regs[30]
